src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java

```diff
@@ -17,156 +17,541 @@
+    @Override
+    public void clearSnapshot(String tag, Map<String, Object> options, String... keyspaceNames)
     {
-        return snapshotLoader.loadSnapshots(keyspace);
+        executeTask(new ClearSnapshotTask(this, getPredicateForCleanedSnapshots(tag, options, keyspaceNames), true));
     }
 
-    public synchronized Set<TableSnapshot> loadSnapshots()
+    @Override
+    public Map<String, TabularData> listSnapshots(Map<String, String> options)
     {
-        return snapshotLoader.loadSnapshots();
+        return new ListSnapshotsTask(this, options, true).call();
     }
 
-    @VisibleForTesting
-    protected synchronized void addSnapshots(Collection<TableSnapshot> snapshots)
+    @Override
+    public long getTrueSnapshotSize()
     {
-        logger.debug("Adding snapshots: {}.", Joiner.on(", ").join(snapshots.stream().map(TableSnapshot::getId).collect(toList())));
-        snapshots.forEach(this::addSnapshot);
+        return new TrueSnapshotSizeTask(this, s -> true).call();
     }
 
-    // TODO: Support pausing snapshot cleanup
-    @VisibleForTesting
-    synchronized void resumeSnapshotCleanup()
+    @Override
+    public long getTrueSnapshotsSize(String keyspace)
     {
-        if (cleanupTaskFuture == null)
+        return new TrueSnapshotSizeTask(this, s -> s.getKeyspaceName().equals(keyspace)).call();
+    }
+
+    @Override
+    public long getTrueSnapshotsSize(String keyspace, String table)
+    {
+        return new TrueSnapshotSizeTask(this, s -> s.getKeyspaceName().equals(keyspace) && s.getTableName().equals(table)).call();
+    }
+
+    @Override
+    public void setSnapshotLinksPerSecond(long throttle)
+    {
+        logger.info("Setting snapshot throttle to {}", throttle);
+        DatabaseDescriptor.setSnapshotLinksPerSecond(throttle);
+    }
+
+    @Override
+    public long getSnapshotLinksPerSecond()
+    {
+        return DatabaseDescriptor.getSnapshotLinksPerSecond();
+    }
+
+    @Override
+    public void handleNotification(INotification notification, Object sender)
+    {
+        if (notification instanceof TruncationNotification)
         {
-            logger.info("Scheduling expired snapshot cleanup with initialDelaySeconds={} and cleanupPeriodSeconds={}",
-                        initialDelaySeconds, cleanupPeriodSeconds);
-            cleanupTaskFuture = executor.scheduleWithFixedDelay(this::clearExpiredSnapshots, initialDelaySeconds,
-                                                                cleanupPeriodSeconds, TimeUnit.SECONDS);
+            TruncationNotification truncationNotification = (TruncationNotification) notification;
+            ColumnFamilyStore cfs = truncationNotification.cfs;
+            if (!truncationNotification.disableSnapshot && cfs.isAutoSnapshotEnabled())
+            {
+                SnapshotOptions opts = SnapshotOptions.systemSnapshot(cfs.name, SnapshotType.TRUNCATE, cfs.getKeyspaceTableName())
```

Comment:
I think we are missing the cfs here.
```suggestion
                SnapshotOptions opts = SnapshotOptions.systemSnapshot(cfs.name, SnapshotType.TRUNCATE, cfs.getKeyspaceTableName()).cfs(cfs)
```

Reply:
@frankgh 

Actually, no ... it is good as it is, for dropping we need to add it, yet, but for truncation we do not. We also do not need that when reacting to `TablePreScrubNotification`. 

The reason we need it only in case of dropping a table is that if we have not passed it there like that, then later on when we are getting CFS in `TakeSnapshotTask`, that CFS would not be found. Look into `dropCf` method in `Keyspace`. There we are removing cfs from `columnFamilyStores` by `tableId`. But later on, when we get CFS in `TakeSnapshotTask`, we would try to get that from the very same map  - which would return null to us. So just for dropping case, we pass just removed cfs into `TableDroppedNotification`.